credentials/xds/xds.go

```diff
@@ -0,0 +1,339 @@
+	isClient bool
+	fallback credentials.TransportCredentials
+}
+
+// handshakeCtxKey is the context key used to store HandshakeInfo values.
+type handshakeCtxKey struct{}
+
+// HandshakeInfo wraps all the security configuration required by client and
+// server handshake methods in credsImpl. The xDS implementation will be
+// responsible for populating these fields.
+//
+// Safe for concurrent access.
+type HandshakeInfo struct {
+	mu               sync.Mutex
+	rootProvider     certprovider.Provider
+	identityProvider certprovider.Provider
+	acceptedSANs     []string // Only on the client side.
+}
+
+// SetRootCertProvider updates the root certificate provider.
+func (chi *HandshakeInfo) SetRootCertProvider(root certprovider.Provider) {
+	chi.mu.Lock()
+	chi.rootProvider = root
+	chi.mu.Unlock()
+}
+
+// SetIdentityCertProvider updates the identity certificate provider.
+func (chi *HandshakeInfo) SetIdentityCertProvider(identity certprovider.Provider) {
+	chi.mu.Lock()
+	chi.identityProvider = identity
+	chi.mu.Unlock()
+}
+
+// SetAcceptedSANs updates the list of accepted SANs.
+func (chi *HandshakeInfo) SetAcceptedSANs(sans []string) {
+	chi.mu.Lock()
+	chi.acceptedSANs = sans
+	chi.mu.Unlock()
+}
+
+func (chi *HandshakeInfo) validate(isClient bool) error {
+	chi.mu.Lock()
+	defer chi.mu.Unlock()
+
+	// On the client side, rootProvider is mandatory. IdentityProvider is
+	// optional based on whether the client is doing TLS or mTLS.
+	if isClient && chi.rootProvider == nil {
+		return errors.New("root certificate provider is missing")
+	}
+
+	// On the server side, identityProvider is mandatory. RootProvider is
+	// optional based on whether the server is doing TLS or mTLS.
+	if !isClient && chi.identityProvider == nil {
+		return errors.New("identity certificate provider is missing")
+	}
+
+	return nil
+}
+
+func (chi *HandshakeInfo) matchingSANExists(cert *x509.Certificate) bool {
+	var sans []string
+	// SANs can be specified in any of these four fields on the parsed cert.
+	sans = append(sans, cert.DNSNames...)
+	sans = append(sans, cert.EmailAddresses...)
+	for _, ip := range cert.IPAddresses {
+		sans = append(sans, ip.String())
+	}
+	for _, uri := range cert.URIs {
+		sans = append(sans, uri.String())
+	}
+
+	chi.mu.Lock()
+	defer chi.mu.Unlock()
+	for _, san := range sans {
+		for _, asan := range chi.acceptedSANs {
+			if san == asan {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// NewHandshakeInfo returns a new instance of HandshakeInfo with the given root
+// and identity certificate providers.
+func NewHandshakeInfo(root, identity certprovider.Provider, sans ...string) *HandshakeInfo {
+	return &HandshakeInfo{
+		rootProvider:     root,
+		identityProvider: identity,
+		acceptedSANs:     sans,
+	}
+}
+
+// NewContextWithHandshakeInfo returns a copy of the parent context with the
+// provided HandshakeInfo stored as a value.
+func NewContextWithHandshakeInfo(parent context.Context, info *HandshakeInfo) context.Context {
```

Comment:
Just curious: can I know the reason why we are storing `HandshakeInfo` in the context instead of directly getting it from `credsImpl`?

Reply:
The provider configurations can be changed anytime by the control plane. These are not specified by the user. Instead when the user uses `XdsCredentials` as their transport credentials, it lets them express their intent in getting security configuration from the control plane. So, whenever new subConns are created the most recent security configuration is added to the context. And when control plane sends new configuration, the context value is mutated by the use of the `Set*()` methods on the `HandshakeInfo` type.

Reply:
Got it. That makes sense. Thanks!